sched/misc/assert.c

```diff
@@ -22,17 +22,150 @@
  * Included Files
  ****************************************************************************/
 
+#include <nuttx/config.h>
+
 #include <nuttx/arch.h>
+#include <nuttx/board.h>
+#include <nuttx/irq.h>
+#include <nuttx/misc/panic.h>
+#include <nuttx/usb/usbdev_trace.h>
+#include <nuttx/syslog/syslog.h>
 
 #include <assert.h>
+#include <debug.h>
 #include <stdlib.h>
 
+#include "irq/irq.h"
+#include "sched/sched.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifndef CONFIG_BOARD_RESET_ON_ASSERT
+#  define CONFIG_BOARD_RESET_ON_ASSERT 0
+#endif
+
+/* USB trace dumping */
+
+#ifndef CONFIG_USBDEV_TRACE
+#  undef CONFIG_ARCH_USBDUMP
+#endif
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: assert_tracecallback
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_USBDUMP
+static int usbtrace_syslog(FAR const char *fmt, ...)
+{
+  va_list ap;
+
+  /* Let vsyslog do the real work */
+
+  va_start(ap, fmt);
+  vsyslog(LOG_EMERG, fmt, ap);
+  va_end(ap);
+  return OK;
+}
+
+static int assert_tracecallback(FAR struct usbtrace_s *trace, FAR void *arg)
+{
+  usbtrace_trprintf(usbtrace_syslog, trace->event, trace->value);
+  return 0;
+}
+#endif
+
+/****************************************************************************
+ * Name: riscv_assert
+ ****************************************************************************/
+
+static void assert_end(void)
+{
+  /* Flush any buffered SYSLOG data */
+
+  syslog_flush();
+
+  /* Are we in an interrupt handler or the idle task? */
+
+  if (up_interrupt_context() || running_task()->flink == NULL)
+    {
+#if CONFIG_BOARD_RESET_ON_ASSERT >= 1
+      board_reset(CONFIG_BOARD_ASSERT_RESET_VALUE);
+#endif
+
+      /* Disable interrupts on this CPU */
+
+      up_irq_save();
+
+#ifdef CONFIG_SMP
+      /* Try (again) to stop activity on other CPUs */
+
+      spin_trylock(&g_cpu_irqlock);
+#endif
+
+      for (; ; )
+        {
+          up_mdelay(250);
+        }
+    }
+  else
+    {
+#if CONFIG_BOARD_RESET_ON_ASSERT >= 2
+      board_reset(CONFIG_BOARD_ASSERT_RESET_VALUE);
+#endif
+    }
+}
+
 /****************************************************************************
  * Public Functions
  ****************************************************************************/
 
 void _assert(FAR const char *filename, int linenum)
 {
+  panic_notifier_call_chain(0, NULL);
```

Comment:
need tell callback whether it's fatal error or not

Reply:
_assert is fatal in my understanding. 
In which case is it needed to tell the difference? I don't get it.

Reply:
No, if the assert is from application, there are option to only abort the current task, not the whole system. Please reference BOARD_RESET_ON_ASSERT:
https://github.com/apache/nuttx/blob/master/boards/Kconfig#L3882-L3891

Reply:
I added some parameter enum, please review